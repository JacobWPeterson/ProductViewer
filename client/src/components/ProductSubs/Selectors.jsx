import React from 'react';
import styled from 'styled-components';

import SizeDropdown from './SizeDropdown.jsx';
import QuantityDropdown from './QuantityDropdown.jsx';

const Wrapper = styled.div`
  width: 100%;
  margin-top: 1vh;
  display: flex;
  flex-direction: row;
`;

class Selectors extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      size: null,
      prevSize: null,
      available: null,
      quantity: null,
      isSizeSelected: false,
      maxQuantity: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
    };
    this.resetThenSet = this.resetThenSet.bind(this);
    this.sizeChangeChecker = this.sizeChangeChecker.bind(this);
  }

  resetThenSet(quantity, size) {
    if (arguments.length === 1) {
      this.setState({
        quantity,
      });
    } else {
      this.setState((prevState) => ({
        prevSize: prevState.size,
        size,
        available: quantity,
        isSizeSelected: !prevState.isSizeSelected,
        headerQuantity: 1,
      }), () => this.sizeChangeChecker());
    }
  }

  sizeChangeChecker() {
    if (this.state.prevSize !== this.state.size) {
      this.setState({ isSizeSelected: true });
    } else {
      this.setState((prevState) => ({ isSizeSelected: true, headerQuantity: prevState.quantity }));
    }
  }

  render() {
    const skus = Object.values(this.props.skus);
    return (
      <Wrapper>
        <SizeDropdown title="SELECT SIZE" list={skus} resetThenSet={this.resetThenSet} />
        {!this.state.isSizeSelected
        && <QuantityDropdown title="â€“" available={this.state.available} list={this.state.maxQuantity} resetThenSet={this.resetThenSet} /> }
        {this.state.isSizeSelected
        && (
        <QuantityDropdown
          title={this.state.headerQuantity}
          available={this.state.available}
          list={this.state.maxQuantity}
          resetThenSet={this.resetThenSet}
        />
        )}
      </Wrapper>
    );
  }
}

export default Selectors;
