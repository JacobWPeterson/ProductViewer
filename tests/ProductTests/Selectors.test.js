/* eslint-disable no-undef */
/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount } from 'enzyme';

import Selectors from '../../client/src/components/ProductSubs/Selectors.jsx';
import ProductTestData from '../ProductTestData.js';

const props = {
  skus: ProductTestData.productStyles.results[0].skus,
};

describe('Selectors', () => {
  const SelectorsComponent = mount(<Selectors {...props} debug />);

  it('Should render correctly', () => {
    expect(SelectorsComponent).toMatchSnapshot();
  });
});
