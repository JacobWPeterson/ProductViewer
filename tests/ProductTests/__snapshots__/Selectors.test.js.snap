// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectors Should render correctly 1`] = `
<Selectors
  debug={true}
  skus={
    Object {
      "407566": Object {
        "quantity": 8,
        "size": "XS",
      },
      "407567": Object {
        "quantity": 16,
        "size": "S",
      },
      "407568": Object {
        "quantity": 17,
        "size": "M",
      },
      "407569": Object {
        "quantity": 10,
        "size": "L",
      },
      "407570": Object {
        "quantity": 15,
        "size": "XL",
      },
      "407571": Object {
        "quantity": 6,
        "size": "XXL",
      },
    }
  }
>
  <Selectors__Wrapper>
    <div
      className="Selectors__Wrapper-sc-1txd6fn-0 hnvKpv"
    >
      <SizeDropdown
        list={
          Array [
            Object {
              "quantity": 8,
              "size": "XS",
            },
            Object {
              "quantity": 16,
              "size": "S",
            },
            Object {
              "quantity": 17,
              "size": "M",
            },
            Object {
              "quantity": 10,
              "size": "L",
            },
            Object {
              "quantity": 15,
              "size": "XL",
            },
            Object {
              "quantity": 6,
              "size": "XXL",
            },
          ]
        }
        resetThenSet={[Function]}
        title="SELECT SIZE"
      >
        <SizeDropdown__DDWrapper>
          <div
            className="SizeDropdown__DDWrapper-u3ohyw-0 knauff"
          >
            <SizeDropdown__DDHeader
              onClick={[Function]}
            >
              <button
                className="SizeDropdown__DDHeader-u3ohyw-1 dFNLhV"
                onClick={[Function]}
              >
                <SizeDropdown__DDHeaderTitle>
                  <div
                    className="SizeDropdown__DDHeaderTitle-u3ohyw-2 fzssNQ"
                  >
                    SELECT SIZE
                  </div>
                </SizeDropdown__DDHeaderTitle>
                <FontAwesome
                  name="angle-down"
                  size="2x"
                >
                  <span
                    aria-hidden={true}
                    className="fa fa-angle-down fa-2x"
                  />
                </FontAwesome>
              </button>
            </SizeDropdown__DDHeader>
          </div>
        </SizeDropdown__DDWrapper>
      </SizeDropdown>
      <QuantityDropdown
        available={null}
        list={
          Array [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
          ]
        }
        resetThenSet={[Function]}
        title="–"
      >
        <QuantityDropdown__DDWrapper>
          <div
            className="QuantityDropdown__DDWrapper-sc-1fwgee9-0 kbjxSC"
          >
            <QuantityDropdown__DDHeader
              onClick={[Function]}
            >
              <button
                className="QuantityDropdown__DDHeader-sc-1fwgee9-1 liwBJT"
                onClick={[Function]}
              >
                <QuantityDropdown__DDHeaderTitle>
                  <div
                    className="QuantityDropdown__DDHeaderTitle-sc-1fwgee9-2 kBnByu"
                  >
                    –
                  </div>
                </QuantityDropdown__DDHeaderTitle>
                <FontAwesome
                  name="angle-down"
                  size="2x"
                >
                  <span
                    aria-hidden={true}
                    className="fa fa-angle-down fa-2x"
                  />
                </FontAwesome>
              </button>
            </QuantityDropdown__DDHeader>
          </div>
        </QuantityDropdown__DDWrapper>
      </QuantityDropdown>
    </div>
  </Selectors__Wrapper>
</Selectors>
`;
